File: internal/manifest/manifest.go
package manifest

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"path"
	"strings"
)

// Manifest describes the files required to bring an installation to a specific version.
type Manifest struct {
	Version string `json:"version"`
	Files   []File `json:"files"`
}

// File represents a single file download entry within a manifest.
type File struct {
	Source string      `json:"src"`
	Target string      `json:"dst"`
	Size   int64       `json:"size"`
	SHA256 string      `json:"sha256"`
	Mode   fs.FileMode `json:"mode,omitempty"`
}

// Parse validates and converts raw JSON manifest bytes into a strongly typed Manifest.
func Parse(data []byte) (Manifest, error) {
	var raw struct {
		Version string            `json:"version"`
		Files   []json.RawMessage `json:"files"`
	}

	if err := json.Unmarshal(data, &raw); err != nil {
		return Manifest{}, fmt.Errorf("manifest: decode: %w", err)
	}

	if strings.TrimSpace(raw.Version) == "" {
		return Manifest{}, errors.New("manifest: version is required")
	}

	if len(raw.Files) == 0 {
		return Manifest{}, errors.New("manifest: files list is empty")
	}

	result := Manifest{Version: raw.Version, Files: make([]File, 0, len(raw.Files))}

	for i, entry := range raw.Files {
		file, err := decodeFile(entry)
		if err != nil {
			return Manifest{}, fmt.Errorf("manifest: files[%d]: %w", i, err)
		}
		result.Files = append(result.Files, file)
	}

	return result, nil
}

func decodeFile(data json.RawMessage) (File, error) {
	var tmp struct {
		Source string  `json:"src"`
		Target string  `json:"dst"`
		Size   *int64  `json:"size"`
		SHA256 string  `json:"sha256"`
		Mode   *string `json:"mode"`
	}

	if err := json.Unmarshal(data, &tmp); err != nil {
		return File{}, fmt.Errorf("decode: %w", err)
	}

	if strings.TrimSpace(tmp.Source) == "" {
		return File{}, errors.New("decode: src is required")
	}

	if path.IsAbs(tmp.Source) {
		return File{}, errors.New("decode: src must be relative")
	}

	if strings.TrimSpace(tmp.Target) == "" {
		return File{}, errors.New("decode: dst is required")
	}

	if path.IsAbs(tmp.Target) {
		return File{}, errors.New("decode: dst must be relative")
	}

	if tmp.Size == nil || *tmp.Size <= 0 {
		return File{}, errors.New("decode: size must be positive")
	}

	if len(tmp.SHA256) != 64 {
		return File{}, errors.New("decode: sha256 must be 64 hex characters")
	}

	if strings.Contains(tmp.Target, "\\") {
		return File{}, errors.New("decode: dst must use forward slashes")
	}

	file := File{
		Source: path.Clean(tmp.Source),
		Target: path.Clean(tmp.Target),
		Size:   *tmp.Size,
		SHA256: strings.ToLower(tmp.SHA256),
		Mode:   0,
	}

	if tmp.Mode != nil && strings.TrimSpace(*tmp.Mode) != "" {
		mode, err := parseFileMode(*tmp.Mode)
		if err != nil {
			return File{}, fmt.Errorf("decode: mode: %w", err)
		}
		file.Mode = mode
	}

	return file, nil
}

func parseFileMode(input string) (fs.FileMode, error) {
	input = strings.TrimSpace(input)
	if len(input) == 0 {
		return 0, errors.New("empty mode")
	}

	var value uint32
	for _, r := range input {
		if r < '0' || r > '7' {
			return 0, fmt.Errorf("invalid digit %q", r)
		}
		value = value*8 + uint32(r-'0')
	}

	return fs.FileMode(value), nil
}

File: src/manifest.rs
use reqwest::Url;
use serde::Deserialize;

use crate::{Error, Result};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Manifest {
    pub version: String,
    pub files: Vec<File>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct File {
    pub source: String,
    pub target: String,
    pub size: u64,
    pub sha256: String,
    pub mode: Option<u32>,
}

#[derive(Deserialize)]
struct RawManifest {
    version: String,
    files: Vec<RawFile>,
}

#[derive(Deserialize)]
struct RawFile {
    #[serde(rename = "src")]
    source: String,
    #[serde(rename = "dst")]
    target: String,
    size: i64,
    sha256: String,
    mode: Option<String>,
}

pub fn parse(data: &[u8]) -> Result<Manifest> {
    let raw: RawManifest = serde_json::from_slice(data)
        .map_err(|error| Error::message(format!("manifest: decode: {error}")))?;

    if raw.version.trim().is_empty() {
        return Err(Error::message("manifest: version is required"));
    }
    if raw.files.is_empty() {
        return Err(Error::message("manifest: files list is empty"));
    }

    let files = raw
        .files
        .into_iter()
        .enumerate()
        .map(|(index, file)| {
            decode_file(file)
                .map_err(|error| Error::message(format!("manifest: files[{index}]: {error}")))
        })
        .collect::<Result<Vec<_>>>()?;

    Ok(Manifest {
        version: raw.version,
        files,
    })
}

fn decode_file(raw: RawFile) -> Result<File> {
    let source = normalize_source(&raw.source)?;
    let target = normalize_target(&raw.target)?;

    if raw.size <= 0 {
        return Err(Error::message("decode: size must be positive"));
    }
    if raw.sha256.len() != 64 || !raw.sha256.bytes().all(|byte| byte.is_ascii_hexdigit()) {
        return Err(Error::message("decode: sha256 must be 64 hex characters"));
    }

    let mode = raw
        .mode
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(parse_file_mode)
        .transpose()?;

    Ok(File {
        source,
        target,
        size: raw.size as u64,
        sha256: raw.sha256.to_ascii_lowercase(),
        mode,
    })
}

fn normalize_source(value: &str) -> Result<String> {
    let value = value.trim();
    if value.is_empty() {
        return Err(Error::message("decode: src is required"));
    }

    if Url::parse(value).is_ok() {
        return Ok(value.to_owned());
    }

    normalize_relative(value, "src")
}

fn normalize_target(value: &str) -> Result<String> {
    let value = value.trim();
    if value.is_empty() {
        return Err(Error::message("decode: dst is required"));
    }
    if value.contains('\\') {
        return Err(Error::message("decode: dst must use forward slashes"));
    }

    normalize_relative(value, "dst")
}

fn normalize_relative(value: &str, field: &str) -> Result<String> {
    if value.starts_with('/') {
        return Err(Error::message(format!("decode: {field} must be relative")));
    }

    let mut parts = Vec::new();
    for part in value.split('/') {
        match part {
            "" | "." => {}
            ".." => {
                if parts.pop().is_none() {
                    return Err(Error::message(format!(
                        "decode: {field} must not escape its root"
                    )));
                }
            }
            part => parts.push(part),
        }
    }

    if parts.is_empty() {
        return Err(Error::message(format!("decode: {field} is required")));
    }

    Ok(parts.join("/"))
}

fn parse_file_mode(value: &str) -> Result<u32> {
    if value.is_empty() {
        return Err(Error::message("decode: mode: empty mode"));
    }
    if !value.bytes().all(|byte| matches!(byte, b'0'..=b'7')) {
        return Err(Error::message(format!(
            "decode: mode: invalid octal mode {value:?}"
        )));
    }

    u32::from_str_radix(value, 8).map_err(|error| Error::message(format!("decode: mode: {error}")))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_valid_manifest() {
        let raw = br#"{
            "version":"1.2.3",
            "files":[{
                "src":"assets/app.bin",
                "dst":"Bin/app.bin",
                "size":42,
                "sha256":"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
                "mode":"0644"
            }]
        }"#;

        let manifest = parse(raw).expect("valid manifest");
        assert_eq!(manifest.version, "1.2.3");
        assert_eq!(manifest.files.len(), 1);
        assert_eq!(manifest.files[0].source, "assets/app.bin");
        assert_eq!(manifest.files[0].target, "Bin/app.bin");
        assert_eq!(manifest.files[0].mode, Some(0o644));
    }

    #[test]
    fn rejects_invalid_entries() {
        let raw = br#"{
            "version":"1.0.0",
            "files":[{"src":"","dst":"file","size":1,"sha256":""}]
        }"#;
        assert!(parse(raw).is_err());
    }

    #[test]
    fn rejects_target_traversal() {
        let raw = br#"{
            "version":"1.0.0",
            "files":[{
                "src":"file.bin",
                "dst":"../outside.bin",
                "size":1,
                "sha256":"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
            }]
        }"#;
        assert!(parse(raw).is_err());
    }
}
